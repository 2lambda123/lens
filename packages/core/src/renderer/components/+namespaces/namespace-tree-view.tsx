import { alpha, SvgIcon, withStyles } from "@material-ui/core";
import { TreeItem, TreeItemProps, TreeView } from "@material-ui/lab";
import { withInjectables } from "@ogre-tools/injectable-react";
import React from "react";
import type { Namespace } from "../../../common/k8s-api/endpoints";
import { DrawerTitle } from "../drawer";
import type { NamespaceStore } from "./store";
import namespaceStoreInjectable from "./store.injectable";

interface NamespaceTreeViewProps {
  root: Namespace;
}

interface Dependencies {
  namespaceStore: NamespaceStore;
}

function NonInjectableNamespaceTreeView({ root, namespaceStore }: Dependencies & NamespaceTreeViewProps) {
  const hierarchicalNamespaces = namespaceStore.getByLabel(["hnc.x-k8s.io/included-namespace=true"]);
  const expandedItems = hierarchicalNamespaces.map(ns => `namespace-${ns.getId()}`);

  function renderBadge(namespace: Namespace) {
    if (!namespace.getAnnotations().find(annotation => annotation.includes("hnc.x-k8s.io/subnamespace-of"))) {
      return null;
    }

    return <span data-testid={`subnamespace-badge-for-${namespace.getId()}`}>S</span>;
  }

  function renderChildren(parent: Namespace) {
    const children = hierarchicalNamespaces.filter(ns =>
      ns.getLabels().find(label => label === `${parent.getName()}.tree.hnc.x-k8s.io/depth=1`)
    );

    return children.map(child => (
      <StyledTreeItem
        key={`namespace-${child.getId()}`}
        nodeId={`namespace-${child.getId()}`}
        data-testid={`namespace-${child.getId()}`}
        label={(
          <>
            {child.getName()}
            {renderBadge(child)}
          </>
        )}
      >
        {renderChildren(child)}
      </StyledTreeItem>
    ));
  }

  if (!root.getLabels().find(label => label === "hnc.x-k8s.io/included-namespace=true")) {
    return null;
  }

  return (
    <div data-testid="namespace-tree-view">
      <DrawerTitle>Tree View</DrawerTitle>
      <TreeView
        defaultExpanded={[`namespace-${root.getId()}`]}
        defaultCollapseIcon={<MinusSquare />}
        defaultExpandIcon={<PlusSquare />}
        defaultEndIcon={<CloseSquare />}
        expanded={expandedItems}
      >
        <StyledTreeItem
          nodeId={`namespace-${root.getId()}`}
          label={root.getName()}
          data-testid={`namespace-${root.getId()}`}
        >
          {renderChildren(root)}
        </StyledTreeItem>
      </TreeView>
    </div>
  )
}

function MinusSquare() {
  return (
    <SvgIcon fontSize="inherit" style={{ width: 14, height: 14 }}>
      <path d="M22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0zM17.873 11.023h-11.826q-.375 0-.669.281t-.294.682v0q0 .401.294 .682t.669.281h11.826q.375 0 .669-.281t.294-.682v0q0-.401-.294-.682t-.669-.281z" />
    </SvgIcon>
  );
}

function PlusSquare() {
  return (
    <SvgIcon fontSize="inherit" style={{ width: 14, height: 14 }}>
      <path d="M22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0zM17.873 12.977h-4.923v4.896q0 .401-.281.682t-.682.281v0q-.375 0-.669-.281t-.294-.682v-4.896h-4.923q-.401 0-.682-.294t-.281-.669v0q0-.401.281-.682t.682-.281h4.923v-4.896q0-.401.294-.682t.669-.281v0q.401 0 .682.281t.281.682v4.896h4.923q.401 0 .682.281t.281.682v0q0 .375-.281.669t-.682.294z" />
    </SvgIcon>
  );
}

function CloseSquare() {
  return (
    <SvgIcon className="close" fontSize="inherit" style={{ width: 14, height: 14 }}>
      <path d="M17.485 17.512q-.281.281-.682.281t-.696-.268l-4.12-4.147-4.12 4.147q-.294.268-.696.268t-.682-.281-.281-.682.294-.669l4.12-4.147-4.12-4.147q-.294-.268-.294-.669t.281-.682.682-.281.696 .268l4.12 4.147 4.12-4.147q.294-.268.696-.268t.682.281 .281.669-.294.682l-4.12 4.147 4.12 4.147q.294.268 .294.669t-.281.682zM22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0z" />
    </SvgIcon>
  );
}

const StyledTreeItem = withStyles((theme) => ({
  iconContainer: {
    '& .close': {
      opacity: 0.3,
    },
  },
  group: {
    marginLeft: 8,
    paddingLeft: 16,
    borderLeft: `1px dashed ${alpha(theme.palette.text.primary, 0.4)}`,
  },
}))((props: TreeItemProps) => <TreeItem {...props} />);

export const NamespaceTreeView = withInjectables<Dependencies, NamespaceTreeViewProps>(NonInjectableNamespaceTreeView, {
  getProps: (di, props) => ({
    namespaceStore: di.inject(namespaceStoreInjectable),
    ...props,
  }),
});
