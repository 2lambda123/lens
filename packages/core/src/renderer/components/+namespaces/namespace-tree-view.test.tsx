import type { AsyncFnMock } from "@async-fn/jest";
import asyncFn from "@async-fn/jest";
import type { DiContainer } from "@ogre-tools/injectable";
import React from "react";
import directoryForKubeConfigsInjectable from "../../../common/app-paths/directory-for-kube-configs/directory-for-kube-configs.injectable";
import directoryForUserDataInjectable from "../../../common/app-paths/directory-for-user-data/directory-for-user-data.injectable";
import type { Fetch } from "../../../common/fetch/fetch.injectable";
import fetchInjectable from "../../../common/fetch/fetch.injectable";
import { Namespace } from "../../../common/k8s-api/endpoints";
import { createMockResponseFromString } from "../../../test-utils/mock-responses";
import hostedClusterInjectable from "../../cluster-frame-context/hosted-cluster.injectable";
import createClusterInjectable from "../../cluster/create-cluster.injectable";
import { getDiForUnitTesting } from "../../getDiForUnitTesting";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import storesAndApisCanBeCreatedInjectable from "../../stores-apis-can-be-created.injectable";
import { type Disposer, disposer } from "../../utils";
import type { DiRender } from "../test-utils/renderFor";
import { renderFor } from "../test-utils/renderFor";
import { NamespaceTreeView } from "./namespace-tree-view";
import type { NamespaceStore } from "./store";
import namespaceStoreInjectable from "./store.injectable";

function createNamespace(name: string, labels?: Record<string, string>, annotations?: Record<string, string>): Namespace {
  return new Namespace({
    apiVersion: "v1",
    kind: "Namespace",
    metadata: {
      name,
      resourceVersion: "1",
      selfLink: `/api/v1/namespaces/${name}`,
      uid: `${name}-1`,
      labels: {
        ...labels
      },
      annotations: {
        ...annotations
      }
    },
  });
}

const singleRoot = createNamespace("single-root", {
  "hnc.x-k8s.io/included-namespace": "true",
});

const acmeGroup = createNamespace("acme-org", {
  "hnc.x-k8s.io/included-namespace": "true",
});

const orgA = createNamespace("org-a", {
  "hnc.x-k8s.io/included-namespace": "true",
});

const teamA = createNamespace("team-a", {
  "hnc.x-k8s.io/included-namespace": "true",
  "acme-org.tree.hnc.x-k8s.io/depth": "1",
  "kubernetes.io/metadata.name": "team-a",
  "team-a.tree.hnc.x-k8s.io/depth": "0",
});

const teamB = createNamespace("team-b", {
  "hnc.x-k8s.io/included-namespace": "true",
  "acme-org.tree.hnc.x-k8s.io/depth": "1",
  "kubernetes.io/metadata.name": "team-b",
  "team-b.tree.hnc.x-k8s.io/depth": "0",
});

const teamC = createNamespace("team-c", {
  "hnc.x-k8s.io/included-namespace": "true",
  "org-a.tree.hnc.x-k8s.io/depth": "1",
  "kubernetes.io/metadata.name": "team-c",
  "team-c.tree.hnc.x-k8s.io/depth": "0",
});

const service1 = createNamespace("service-1", {
  "hnc.x-k8s.io/included-namespace": "true",
  "org-a.tree.hnc.x-k8s.io/depth": "1",
  "kubernetes.io/metadata.name": "team-c",
  "service-1.tree.hnc.x-k8s.io/depth": "0",
}, {
  "hnc.x-k8s.io/subnamespace-of": "org-a",
});

describe("<NamespaceTreeView />", () => {
  let di: DiContainer;
  let render: DiRender;
  let namespaceStore: NamespaceStore;
  let fetchMock: AsyncFnMock<Fetch>;
  let cleanup: Disposer;
  
  beforeEach(async () => {
    di = getDiForUnitTesting({ doGeneralOverrides: true });
    di.unoverride(subscribeStoresInjectable);

    di.override(directoryForUserDataInjectable, () => "/some-user-store-path");
    di.override(directoryForKubeConfigsInjectable, () => "/some-kube-configs");
    di.override(storesAndApisCanBeCreatedInjectable, () => true);

    fetchMock = asyncFn();
    di.override(fetchInjectable, () => fetchMock);

    const createCluster = di.inject(createClusterInjectable);

    di.override(hostedClusterInjectable, () => createCluster({
      contextName: "some-context-name",
      id: "some-cluster-id",
      kubeConfigPath: "/some-path-to-a-kubeconfig",
    }, {
      clusterServerUrl: "https://localhost:8080",
    }));

    namespaceStore = di.inject(namespaceStoreInjectable);

    const subscribeStores = di.inject(subscribeStoresInjectable);

    cleanup = disposer(subscribeStores([namespaceStore]));
    render = renderFor(di);
  });

  afterEach(() => {
    cleanup();
  });

  describe("once the subscribe resolves", () => {
    beforeEach(async () => {
      await fetchMock.resolveSpecific([
        "https://127.0.0.1:12345/api-kube/api/v1/namespaces",
      ], createMockResponseFromString("https://127.0.0.1:12345/api-kube/api/v1/namespaces", JSON.stringify({
        apiVersion: "v1",
        kind: "NamespaceList",
        metadata: {},
        items: [
          createNamespace("test-1"),
          createNamespace("test-2"),
          createNamespace("test-3"),
          createNamespace("test-4"),
          createNamespace("test-5"),
          acmeGroup,
          orgA,
          teamA,
          teamB,
          teamC,
          service1,
        ],
      })));
    });

    it("renders null with regular namespace", () => {
      const result = render(<NamespaceTreeView root={createNamespace("tree-1")} />);
  
      expect(result.baseElement).toMatchSnapshot();
    });

    it("renders one namespace without children", () => {
      const result = render(<NamespaceTreeView root={singleRoot} />);
  
      expect(result.baseElement).toMatchSnapshot();
    });

    it("renders namespace with 2 children namespaces", () => {
      const result = render(<NamespaceTreeView root={acmeGroup} />);

      expect(result.baseElement).toMatchSnapshot();
    });

    it("renders namespace with children namespaces and a subnamespace", () => {
      const result = render(<NamespaceTreeView root={orgA} />);

      expect(result.baseElement).toMatchSnapshot();
    });

    it("renders an indicator badge for the subnamespace", () => {
      const result = render(<NamespaceTreeView root={orgA} />);

      expect(result.getByTestId("subnamespace-badge-for-service-1-1")).toBeInTheDocument();
    });

    it("does not render an indicator badge for the true namespace", () => {
      const result = render(<NamespaceTreeView root={orgA} />);
      const trueNamespace = result.getByTestId("namespace-team-c-1");

      expect(trueNamespace.querySelector("[data-testid='subnamespace-badge-for-team-c-1']")).toBeNull();
    });
  });
});