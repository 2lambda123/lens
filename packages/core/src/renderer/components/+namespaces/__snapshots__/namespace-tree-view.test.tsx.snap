// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NamespaceTreeView /> once the subscribe resolves renders 2 levels deep 1`] = `
<body>
  <div>
    <div
      data-testid="namespace-tree-view"
    >
      <div
        class="DrawerTitle title"
      >
        Tree View
      </div>
      <ul
        aria-multiselectable="false"
        class="MuiTreeView-root"
        role="tree"
      >
        <li
          aria-expanded="true"
          class="MuiTreeItem-root Mui-expanded"
          data-testid="namespace-levels-deep-1"
          role="treeitem"
          tabindex="0"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer Component-iconContainer-11"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                style="width: 14px; height: 14px;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0zM17.873 11.023h-11.826q-.375 0-.669.281t-.294.682v0q0 .401.294 .682t.669.281h11.826q.375 0 .669-.281t.294-.682v0q0-.401-.294-.682t-.669-.281z"
                />
              </svg>
            </div>
            <div
              class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
            >
              levels-deep
            </div>
          </div>
          <ul
            class="MuiCollapse-root MuiTreeItem-group Component-group-12 MuiCollapse-entered"
            role="group"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner"
              >
                <li
                  class="MuiTreeItem-root Mui-expanded"
                  data-testid="namespace-level-deep-child-a-1"
                  role="treeitem"
                  tabindex="-1"
                >
                  <div
                    class="MuiTreeItem-content"
                  >
                    <div
                      class="MuiTreeItem-iconContainer Component-iconContainer-11"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root close MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        style="width: 14px; height: 14px;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M17.485 17.512q-.281.281-.682.281t-.696-.268l-4.12-4.147-4.12 4.147q-.294.268-.696.268t-.682-.281-.281-.682.294-.669l4.12-4.147-4.12-4.147q-.294-.268-.294-.669t.281-.682.682-.281.696 .268l4.12 4.147 4.12-4.147q.294-.268.696-.268t.682.281 .281.669-.294.682l-4.12 4.147 4.12 4.147q.294.268 .294.669t-.281.682zM22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0z"
                        />
                      </svg>
                    </div>
                    <div
                      class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                    >
                      level-deep-child-a
                    </div>
                  </div>
                  <ul
                    class="MuiCollapse-root MuiTreeItem-group Component-group-12 MuiCollapse-entered"
                    role="group"
                    style="min-height: 0px;"
                  >
                    <div
                      class="MuiCollapse-wrapper"
                    >
                      <div
                        class="MuiCollapse-wrapperInner"
                      />
                    </div>
                  </ul>
                </li>
                <li
                  aria-expanded="true"
                  class="MuiTreeItem-root Mui-expanded"
                  data-testid="namespace-level-deep-child-b-1"
                  role="treeitem"
                  tabindex="-1"
                >
                  <div
                    class="MuiTreeItem-content"
                  >
                    <div
                      class="MuiTreeItem-iconContainer Component-iconContainer-11"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        style="width: 14px; height: 14px;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0zM17.873 11.023h-11.826q-.375 0-.669.281t-.294.682v0q0 .401.294 .682t.669.281h11.826q.375 0 .669-.281t.294-.682v0q0-.401-.294-.682t-.669-.281z"
                        />
                      </svg>
                    </div>
                    <div
                      class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                    >
                      level-deep-child-b
                    </div>
                  </div>
                  <ul
                    class="MuiCollapse-root MuiTreeItem-group Component-group-12 MuiCollapse-entered"
                    role="group"
                    style="min-height: 0px;"
                  >
                    <div
                      class="MuiCollapse-wrapper"
                    >
                      <div
                        class="MuiCollapse-wrapperInner"
                      >
                        <li
                          class="MuiTreeItem-root Mui-expanded"
                          data-testid="namespace-level-deep-subchild-a-1"
                          role="treeitem"
                          tabindex="-1"
                        >
                          <div
                            class="MuiTreeItem-content"
                          >
                            <div
                              class="MuiTreeItem-iconContainer Component-iconContainer-11"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root close MuiSvgIcon-fontSizeInherit"
                                focusable="false"
                                style="width: 14px; height: 14px;"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M17.485 17.512q-.281.281-.682.281t-.696-.268l-4.12-4.147-4.12 4.147q-.294.268-.696.268t-.682-.281-.281-.682.294-.669l4.12-4.147-4.12-4.147q-.294-.268-.294-.669t.281-.682.682-.281.696 .268l4.12 4.147 4.12-4.147q.294-.268.696-.268t.682.281 .281.669-.294.682l-4.12 4.147 4.12 4.147q.294.268 .294.669t-.281.682zM22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0z"
                                />
                              </svg>
                            </div>
                            <div
                              class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                            >
                              level-deep-subchild-a
                            </div>
                          </div>
                          <ul
                            class="MuiCollapse-root MuiTreeItem-group Component-group-12 MuiCollapse-entered"
                            role="group"
                            style="min-height: 0px;"
                          >
                            <div
                              class="MuiCollapse-wrapper"
                            >
                              <div
                                class="MuiCollapse-wrapperInner"
                              />
                            </div>
                          </ul>
                        </li>
                      </div>
                    </div>
                  </ul>
                </li>
              </div>
            </div>
          </ul>
        </li>
      </ul>
    </div>
  </div>
</body>
`;

exports[`<NamespaceTreeView /> once the subscribe resolves renders namespace with 2 children namespaces 1`] = `
<body>
  <div>
    <div
      data-testid="namespace-tree-view"
    >
      <div
        class="DrawerTitle title"
      >
        Tree View
      </div>
      <ul
        aria-multiselectable="false"
        class="MuiTreeView-root"
        role="tree"
      >
        <li
          aria-expanded="true"
          class="MuiTreeItem-root Mui-expanded"
          data-testid="namespace-acme-org-1"
          role="treeitem"
          tabindex="0"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer Component-iconContainer-3"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                style="width: 14px; height: 14px;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0zM17.873 11.023h-11.826q-.375 0-.669.281t-.294.682v0q0 .401.294 .682t.669.281h11.826q.375 0 .669-.281t.294-.682v0q0-.401-.294-.682t-.669-.281z"
                />
              </svg>
            </div>
            <div
              class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
            >
              acme-org
            </div>
          </div>
          <ul
            class="MuiCollapse-root MuiTreeItem-group Component-group-4 MuiCollapse-entered"
            role="group"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner"
              >
                <li
                  class="MuiTreeItem-root Mui-expanded"
                  data-testid="namespace-team-a-1"
                  role="treeitem"
                  tabindex="-1"
                >
                  <div
                    class="MuiTreeItem-content"
                  >
                    <div
                      class="MuiTreeItem-iconContainer Component-iconContainer-3"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root close MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        style="width: 14px; height: 14px;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M17.485 17.512q-.281.281-.682.281t-.696-.268l-4.12-4.147-4.12 4.147q-.294.268-.696.268t-.682-.281-.281-.682.294-.669l4.12-4.147-4.12-4.147q-.294-.268-.294-.669t.281-.682.682-.281.696 .268l4.12 4.147 4.12-4.147q.294-.268.696-.268t.682.281 .281.669-.294.682l-4.12 4.147 4.12 4.147q.294.268 .294.669t-.281.682zM22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0z"
                        />
                      </svg>
                    </div>
                    <div
                      class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                    >
                      team-a
                    </div>
                  </div>
                  <ul
                    class="MuiCollapse-root MuiTreeItem-group Component-group-4 MuiCollapse-entered"
                    role="group"
                    style="min-height: 0px;"
                  >
                    <div
                      class="MuiCollapse-wrapper"
                    >
                      <div
                        class="MuiCollapse-wrapperInner"
                      />
                    </div>
                  </ul>
                </li>
                <li
                  class="MuiTreeItem-root Mui-expanded"
                  data-testid="namespace-team-b-1"
                  role="treeitem"
                  tabindex="-1"
                >
                  <div
                    class="MuiTreeItem-content"
                  >
                    <div
                      class="MuiTreeItem-iconContainer Component-iconContainer-3"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root close MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        style="width: 14px; height: 14px;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M17.485 17.512q-.281.281-.682.281t-.696-.268l-4.12-4.147-4.12 4.147q-.294.268-.696.268t-.682-.281-.281-.682.294-.669l4.12-4.147-4.12-4.147q-.294-.268-.294-.669t.281-.682.682-.281.696 .268l4.12 4.147 4.12-4.147q.294-.268.696-.268t.682.281 .281.669-.294.682l-4.12 4.147 4.12 4.147q.294.268 .294.669t-.281.682zM22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0z"
                        />
                      </svg>
                    </div>
                    <div
                      class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                    >
                      team-b
                    </div>
                  </div>
                  <ul
                    class="MuiCollapse-root MuiTreeItem-group Component-group-4 MuiCollapse-entered"
                    role="group"
                    style="min-height: 0px;"
                  >
                    <div
                      class="MuiCollapse-wrapper"
                    >
                      <div
                        class="MuiCollapse-wrapperInner"
                      />
                    </div>
                  </ul>
                </li>
              </div>
            </div>
          </ul>
        </li>
      </ul>
    </div>
  </div>
</body>
`;

exports[`<NamespaceTreeView /> once the subscribe resolves renders namespace with children namespaces and a subnamespace 1`] = `
<body>
  <div>
    <div
      data-testid="namespace-tree-view"
    >
      <div
        class="DrawerTitle title"
      >
        Tree View
      </div>
      <ul
        aria-multiselectable="false"
        class="MuiTreeView-root"
        role="tree"
      >
        <li
          aria-expanded="true"
          class="MuiTreeItem-root Mui-expanded"
          data-testid="namespace-org-a-1"
          role="treeitem"
          tabindex="0"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer Component-iconContainer-5"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                style="width: 14px; height: 14px;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0zM17.873 11.023h-11.826q-.375 0-.669.281t-.294.682v0q0 .401.294 .682t.669.281h11.826q.375 0 .669-.281t.294-.682v0q0-.401-.294-.682t-.669-.281z"
                />
              </svg>
            </div>
            <div
              class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
            >
              org-a
            </div>
          </div>
          <ul
            class="MuiCollapse-root MuiTreeItem-group Component-group-6 MuiCollapse-entered"
            role="group"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner"
              >
                <li
                  class="MuiTreeItem-root Mui-expanded"
                  data-testid="namespace-service-1-1"
                  role="treeitem"
                  tabindex="-1"
                >
                  <div
                    class="MuiTreeItem-content"
                  >
                    <div
                      class="MuiTreeItem-iconContainer Component-iconContainer-5"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root close MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        style="width: 14px; height: 14px;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M17.485 17.512q-.281.281-.682.281t-.696-.268l-4.12-4.147-4.12 4.147q-.294.268-.696.268t-.682-.281-.281-.682.294-.669l4.12-4.147-4.12-4.147q-.294-.268-.294-.669t.281-.682.682-.281.696 .268l4.12 4.147 4.12-4.147q.294-.268.696-.268t.682.281 .281.669-.294.682l-4.12 4.147 4.12 4.147q.294.268 .294.669t-.281.682zM22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0z"
                        />
                      </svg>
                    </div>
                    <div
                      class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                    >
                      service-1
                      <span
                        data-testid="subnamespace-badge-for-service-1-1"
                      >
                        S
                      </span>
                    </div>
                  </div>
                  <ul
                    class="MuiCollapse-root MuiTreeItem-group Component-group-6 MuiCollapse-entered"
                    role="group"
                    style="min-height: 0px;"
                  >
                    <div
                      class="MuiCollapse-wrapper"
                    >
                      <div
                        class="MuiCollapse-wrapperInner"
                      />
                    </div>
                  </ul>
                </li>
                <li
                  class="MuiTreeItem-root Mui-expanded"
                  data-testid="namespace-team-c-1"
                  role="treeitem"
                  tabindex="-1"
                >
                  <div
                    class="MuiTreeItem-content"
                  >
                    <div
                      class="MuiTreeItem-iconContainer Component-iconContainer-5"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root close MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        style="width: 14px; height: 14px;"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M17.485 17.512q-.281.281-.682.281t-.696-.268l-4.12-4.147-4.12 4.147q-.294.268-.696.268t-.682-.281-.281-.682.294-.669l4.12-4.147-4.12-4.147q-.294-.268-.294-.669t.281-.682.682-.281.696 .268l4.12 4.147 4.12-4.147q.294-.268.696-.268t.682.281 .281.669-.294.682l-4.12 4.147 4.12 4.147q.294.268 .294.669t-.281.682zM22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0z"
                        />
                      </svg>
                    </div>
                    <div
                      class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
                    >
                      team-c
                    </div>
                  </div>
                  <ul
                    class="MuiCollapse-root MuiTreeItem-group Component-group-6 MuiCollapse-entered"
                    role="group"
                    style="min-height: 0px;"
                  >
                    <div
                      class="MuiCollapse-wrapper"
                    >
                      <div
                        class="MuiCollapse-wrapperInner"
                      />
                    </div>
                  </ul>
                </li>
              </div>
            </div>
          </ul>
        </li>
      </ul>
    </div>
  </div>
</body>
`;

exports[`<NamespaceTreeView /> once the subscribe resolves renders null with regular namespace 1`] = `
<body>
  <div />
</body>
`;

exports[`<NamespaceTreeView /> once the subscribe resolves renders one namespace without children 1`] = `
<body>
  <div>
    <div
      data-testid="namespace-tree-view"
    >
      <div
        class="DrawerTitle title"
      >
        Tree View
      </div>
      <ul
        aria-multiselectable="false"
        class="MuiTreeView-root"
        role="tree"
      >
        <li
          class="MuiTreeItem-root"
          data-testid="namespace-single-root-1"
          role="treeitem"
          tabindex="0"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer Component-iconContainer-1"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root close MuiSvgIcon-fontSizeInherit"
                focusable="false"
                style="width: 14px; height: 14px;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17.485 17.512q-.281.281-.682.281t-.696-.268l-4.12-4.147-4.12 4.147q-.294.268-.696.268t-.682-.281-.281-.682.294-.669l4.12-4.147-4.12-4.147q-.294-.268-.294-.669t.281-.682.682-.281.696 .268l4.12 4.147 4.12-4.147q.294-.268.696-.268t.682.281 .281.669-.294.682l-4.12 4.147 4.12 4.147q.294.268 .294.669t-.281.682zM22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0z"
                />
              </svg>
            </div>
            <div
              class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
            >
              single-root
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</body>
`;
